cartographer/mapping/pose_extrapolator_test.cc

```diff
@@ -0,0 +1,186 @@
+constexpr double kGravityTimeConstant = 0.1f;
+constexpr double kExtrapolateDuration = 0.1f;
+constexpr double kPrecision = 1e-8;
+constexpr double kExtrapolatePrecision = 1e-2;
+
+TEST(PoseExtrapolatorDeathTest, IncompleteInitialization) {
+  PoseExtrapolator extrapolator(common::FromSeconds(kPoseQueueDuration),
+                                kGravityTimeConstant);
+  common::Time time = common::FromUniversal(123);
+  EXPECT_DEATH(extrapolator.EstimateGravityOrientation(time), "");
+  EXPECT_DEATH(extrapolator.ExtrapolatePose(time), "");
+  Eigen::Vector3d acceleration(0, 0, 9);
+  Eigen::Vector3d angular_velocity(0, 0, 0);
+  extrapolator.AddImuData(
+      sensor::ImuData{time, acceleration, angular_velocity});
+  EXPECT_DEATH(extrapolator.EstimateGravityOrientation(time), "");
+  EXPECT_DEATH(extrapolator.ExtrapolatePose(time), "");
+}
+
+TEST(PoseExtrapolatorDeathTest, ExtrapolateInPast) {
+  PoseExtrapolator extrapolator(common::FromSeconds(kPoseQueueDuration),
+                                kGravityTimeConstant);
+  common::Time time_present = common::FromUniversal(123);
+  transform::Rigid3d pose =
+      transform::Rigid3d::Translation(Eigen::Vector3d(0.1, 0.2, 0.3));
+  extrapolator.AddPose(time_present, pose);
+  extrapolator.EstimateGravityOrientation(time_present);
+  EXPECT_THAT(extrapolator.ExtrapolatePose(time_present),
+              transform::IsNearly(pose, kPrecision));
+  common::Time time_in_past = time_present - common::FromSeconds(10);
+  EXPECT_DEATH(extrapolator.EstimateGravityOrientation(time_in_past), "");
+  EXPECT_DEATH(extrapolator.ExtrapolatePose(time_in_past), "");
+  common::Time time_in_future = time_present + common::FromSeconds(20);
+  extrapolator.ExtrapolatePose(time_in_future);
+  EXPECT_DEATH(extrapolator.ExtrapolatePose(time_present), "");
+}
+
+TEST(PoseExtrapolatorTest, EstimateGravityOrientationWithIMU) {
+  Eigen::Vector3d initial_gravity_acceleration(1.6, 2.0, 8.0);
+  Eigen::Vector3d angular_velocity(0, 0, 0);
+  common::Time current_time = common::FromUniversal(123);
+  sensor::ImuData imu_data{current_time, initial_gravity_acceleration,
+                           angular_velocity};
+  auto extrapolator = PoseExtrapolator::InitializeWithImu(
+      common::FromSeconds(kPoseQueueDuration), kGravityTimeConstant, imu_data);
+  {
+    Eigen::Quaterniond expected_orientation;
+    expected_orientation.setFromTwoVectors(initial_gravity_acceleration,
+                                           Eigen::Vector3d::UnitZ());
+    EXPECT_NEAR(0.,
+                extrapolator->EstimateGravityOrientation(current_time)
+                    .angularDistance(expected_orientation),
+                kPrecision);
+  }
+  Eigen::Vector3d gravity_acceleration(1.6, 2.0, 8.0);
+  for (int i = 0; i < 10; ++i) {
+    current_time += common::FromSeconds(kGravityTimeConstant);
+    extrapolator->AddImuData(
+        sensor::ImuData{current_time, gravity_acceleration, angular_velocity});
+  }
+  {
```

Comment:
You do not have create a scope for every vector/quaternion.

Reply:
Done